### ISSUE TITLE
Could the buffer list be shown faster?

### ISSUE THREAD

**Issue by x-ji** [tridactyl/tridactyl#1242]
In Vimium the buffer list shows up almost instantly, no matter how many tabs I have. In Tridactyl there's usually a delay of more than one second. Therefore if I try to type a search term before the list is actually shown, unexpected behaviors happen.

The buffer list is the biggest pain point I have with Tridactyl, both the speed and the order of the tabs (#1040).

**Comment by bovine3dom** [comment]
I think this is a deeper architectural problem. If you type `on` very quickly, the `n` key gets lost (true for any key after the `o`, obviously).

I'm not sure how we can fix this short of stopping all pretense of the command line being a real text box and instead creating our own queue of keys that should appear in it. That could well make apparent latency appear even worse, but it would have the benefit of always doing things in the right order.

Something to think about for #932.

**Comment by arcnmx** [comment]
Is that is the same problem? In my experience the commandline appears quickly enough, and typing in it is fine, it's simply the completion list of open tabs that takes an unreasonably long time to populate.

**Comment by bovine3dom** [comment]
It's somewhat regards but probably closer to #1208.

**Comment by x-ji** [comment]
Yeah the mistyping of the commands is a problem but the slow populating of the list also makes the functionality largely unusable. I'd rather look at my sidebar to find the tab instead of waiting for more than 1 second.

**Comment by glacambre** [comment]
@x-ji If you're on the betas, completions should be slightly faster to pop up with the latest version.

**Comment by dkasak** [comment]
I can confirm that this is most likely a separate issue because it often takes 10+ seconds for the buffer list to appear for me. I have a lot of open tabs but not enough that it should realistically cause such a big slowdown (say 100-200).

**Comment by glacambre** [comment]
@dkasak If you're using the betas, could you try updating (about:addons -> cog icon -> check for updates) and telling us if things are better? We made a few small changes today that should help with this issue.

**Comment by dkasak** [comment]
I wasn't using the betas, but I tried it out now.

Using 1.14.8-133-ge3ae6f8, pressing `b` almost immediately responds by showing the empty buffer list (grey header line saying `Tabs` and a white empty area where the tabs are supposed to be listed). Previously only the `Tabs` header line was shown and I think it was a bit laggier. However, it still takes a long time to partially populate, though it's quicker than before (~2 seconds).

While the list is partially populated, there is no prompt at the bottom of the buffer list. The prompt (and the fully populated list) still takes a lot of time to show up (anything between 6 and 10 seconds). It feels like this is the same duration as before, only now it displays a partial result earlier. However, if I try to interrupt it before it is finished (i.e. before the prompt is shown) by pressing `Escape`, it stalls until the time the prompt would be shown and only then closes the buffer list. During that time, the whole browser UI is frozen.

**Comment by mozbugbox** [comment]
It seems this line took tons of time in a for loop:

https://github.com/tridactyl/tridactyl/blob/d813b09ae44c1389b365e42baf71fb29361a9b5a/src/completions/Tab.ts#L168

async call in a for loop ...

Once replaced the container lookup with a constant, the completions for `tab` pop up much quicker, though I can still see a small delay.

No idea what the  container was for.

**Comment by bovine3dom** [comment]
It shows the icon of the container (e.g. "Banking") a tab is in. Presumably it'd be N times quicker if we moved it out of the for loop and put it in an `await Promise.all(...`.

**Comment by glacambre** [comment]
@mozbugbox Nice find! Containers are firefox containers, so we could probably replace the getFromId call with `browser.contextualIdentities.query()` to get a list of all contextual identities and then synchronously search for identities in the returned array.

**Comment by notpentadactyl** [comment]
any update on this?

**Comment by bovine3dom** [comment]
#2353 may have sped it up a little. Otherwise our main lead is that `Tab.ts` line mentioned above. No one has investigated it yet.

**Comment by bovine3dom** [comment]
https://github.com/tridactyl/tridactyl/commit/ba9cce3e371dbaeb54285e00190e7026f6201744 should have sped it up quite a bit too.

**Comment by notpentadactyl** [comment]
I have around 650 tabs. Pressing `b` takes over 30 seconds to populate and be responsive again.
It also makes my CPU heat up and drives the fan spinning at full speed.
This happens every time I press `b`, doesn't matter if I just did it.

**Comment by bovine3dom** [comment]
That's quite amazing. What version of Tridactyl are you using? `:version`.

**Comment by notpentadactyl** [comment]
> 
> 
> That's quite amazing. What version of Tridactyl are you using? `:version`.

`https://twitter.com/CNN/status/1266323257520766976`

**Comment by notpentadactyl** [comment]
> 
> 
> That's quite amazing. What version of Tridactyl are you using? `:version`.

`1.19.0pre3923-dda987d`

**Comment by bovine3dom** [comment]
I can't reproduce this with 700 tabs in a disposable profile:

https://gfycat.com/glisteningglaringeuropeanfiresalamander

It does take a second to show. Presumably we can improve that.

I also tried with `:set tabsort mru` and had the same outcome.

**Comment by notpentadactyl** [comment]
@bovine3dom I debugged the issue and found the culprit: the extension [decentraleyes](https://decentraleyes.org/). 

It now takes ~1 second to show. Thanks!

**Comment by dkasak** [comment]
I can confirm that turning off DecentralEyes makes my problem much better as well.
